taichi/lang_util.cpp

```diff
@@ -29,6 +29,37 @@ real get_cpu_frequency() {
 
 real default_measurement_time = 1;
 
+// Note: these primitive types should never be freed. They are supposed to live
+// together with the process.
+#define PER_TYPE(x)      \
+  DataType DataType::x = \
+      DataType(new PrimitiveType(PrimitiveType::primitive_type::x));
+#include "taichi/inc/data_type.inc.h"
+#undef PER_TYPE
+
+DataType PrimitiveType::get(PrimitiveType::primitive_type t) {
+  if (false) {
+  }
+#define PER_TYPE(x) else if (t == primitive_type::x) return DataType::x;
```

Comment:
nit: this creates a coupling between `PrimitiveType::x` and `DataType::x`. If it is `PrimitiveType` that maintains those global pointers, and knows how to map from `PrimitiveType::primitive_type` to the pointer, then we can just return `DataType(map_to_ptr(t))`?

Reply:
Good point. In fact, the next step I would do is to replace all `DataType::x` with `PrimitiveType` using @taichi-gardener. That would be a huge change, and I wouldn't want that to mix with what we have in this PR. Hence the temporary solution here :-)